networking_cisco/plugins/cisco/cfg_agent/device_drivers/asr1k/aci_asr1k_routing_driver.py

```diff
@@ -230,19 +251,60 @@ def cleanup_invalid_cfg(self, hd, routers):
         cfg_syncer = syncer.ConfigSyncer(routers, self, hd)
         cfg_syncer.delete_invalid_cfg()
 
+    def _create_vrf(self, ri):
+        vrf_name = self._get_vrf_name(ri)
+        # Don't create the VRF if we don't have a
+        # name yet -- it will be deferred to when
+        # have one
+        if vrf_name:
+            self._do_create_vrf(vrf_name)
+
+    def _remove_vrf(self, ri):
+        vrf_name = self._get_vrf_name(ri)
+        if vrf_name:
+            self._do_remove_vrf(vrf_name)
+
+    def _do_create_vrf(self, vrf_name):
+        conf_str = csr_snippets.CREATE_VRF % vrf_name
+        if not self._cfg_exists(conf_str):
+            self._edit_running_config(conf_str, 'CREATE_VRF')
+
     def _get_vrf_name(self, ri):
         """
-        For ACI, a tenant is mapped to a VRF.
+        For ACI, there are different mappings for VRFs. The
+        VLAN used on the internal interfaces of a neutron
+        router represent the VRF (there are modes where
+        different neutron routers will have interfaces
+        in the same VRF, which is why the VLANs are used
+        to get the mapping). If it's the global router,
+        the gateway port's VLAN will be used instead
+        of an interface port.
         """
-        router_id = (helper.N_ROUTER_PREFIX +
-                     ri.router['tenant_id'])[:self.DEV_NAME_LEN]
+        vrf_name = ri.router.get('vrf_name')
+        if vrf_name:
+            return vrf_name
+        # haven't created it yet, so go create it
+        interfaces = ri.router.get(constants.INTERFACE_KEY, [])
+        if not interfaces:
+            return None
+        # If the apic_system_id property is set, then we know
```

Comment:
the comment mentioned apic_system_id which is not used anywhere in the code below.....This comment is confusing to me here.....